src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyIconSetCache.cs

```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using Microsoft.VisualStudio.Imaging.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Dependencies
+{
+    /// <summary>
+    /// Caches and reuses <see cref="DependencyIconSet"/> instances.
+    /// In practice dependencies use a relatively small number of distinct icon
+    /// sets; we can save considerable amounts of memory by ensuring that the same
+    /// logical sets are represented with the same instances.
+    /// </summary>
+    internal sealed class DependencyIconSetCache
+    {
+        private readonly ConcurrentDictionary<DependencyIconSet, DependencyIconSet> _iconSets = new ConcurrentDictionary<DependencyIconSet, DependencyIconSet>();
```

Comment:
Nit: ThreadingTools.ApplyChangeOptimistically + ImmutableHashSet provides a nice lock-free implementation of this, see https://github.com/dotnet/project-system/pull/3960/files#diff-0e5ddc3ccc529a971c09299d53572076R67 for an example.

Reply:
@davkean I'll implement that. Thanks!